internal/muxado/buffer.go

```diff
@@ -77,30 +119,34 @@ func (b *inboundBuffer) Read(p []byte) (n int, err error) {
 func (b *inboundBuffer) SetError(err error) {
 	b.mu.Lock()
 	b.err = err
-	b.mu.Unlock()
 	b.cond.Broadcast()
+	b.mu.Unlock()
 }
 
 func (b *inboundBuffer) SetDeadline(t time.Time) {
-	// XXX: implement
-	/*
-		b.L.Lock()
-
-		// set the deadline
-		b.deadline = t
+	b.mu.Lock()
 
-		// how long until the deadline
-		delay := t.Sub(time.Now())
+	// Set the deadline and notify any readers that they need to take heed.
+	// They'll figure out all of the timer management for us.
+	b.deadline = t
+	if timeout := time.Until(t); timeout > 0 {
+		b.startTimerLocked(timeout)
+	}
```

Comment:
maybe a bit nicer would be to stop the timer as an else here 🤔

Reply:
and then `Broadcast` all the time 🤷

Reply:
I think we end up doing an unneeded broadcast on deadline == zero, but that's probably ok :shrug: